contracts/swaps/connectors/SwapsImplstETH_ETH.sol

```diff
@@ -0,0 +1,175 @@
+        address sourceTokenAddress,
+        address destTokenAddress,
+        address receiverAddress,
+        address returnToSenderAddress,
+        uint256 minSourceTokenAmount,
+        uint256 maxSourceTokenAmount,
+        uint256 requiredDestTokenAmount,
+        bytes memory payload
+    )
+        public
+        returns (uint256 destTokenAmountReceived, uint256 sourceTokenAmountUsed)
+    {
+        require(sourceTokenAddress != destTokenAddress, "source == dest");
+        require(
+            (sourceTokenAddress == WSTETH || sourceTokenAddress == WETH) &&
+                (destTokenAddress == WSTETH || destTokenAddress == WETH),
+            "unsupported tokens"
+        );
+
+        IERC20 sourceToken = IERC20(sourceTokenAddress);
+        address _thisAddress = address(this);
+        (sourceTokenAmountUsed, destTokenAmountReceived) = _swap(
+            sourceTokenAddress,
+            destTokenAddress,
+            receiverAddress,
+            minSourceTokenAmount,
+            maxSourceTokenAmount,
+            requiredDestTokenAmount,
+            payload
+        );
+
+        if (
+            returnToSenderAddress != _thisAddress &&
+            sourceTokenAmountUsed < maxSourceTokenAmount
+        ) {
+            // send unused source token back
+            sourceToken.safeTransfer(
+                returnToSenderAddress,
+                maxSourceTokenAmount - sourceTokenAmountUsed
+            );
+        }
+    }
+
+    function dexExpectedRate(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        uint256 sourceTokenAmount
+    ) public view returns (uint256 expectedRate) {
+        revert("unsupported");
+    }
+
+    function dexAmountOut(bytes memory payload, uint256 amountIn)
+        public
+        returns (uint256 amountOut, address midToken)
+    {
+        if (amountIn != 0) {
+            (, address srcToken, address destToken) = abi.decode(
+                payload,
+                (uint256, address, address)
+            );
+            if (srcToken == WETH) {
+                amountOut = IwstETH(WSTETH).getWstETHBystETH(amountIn);
+            } else {
+                amountOut = IwstETH(WSTETH).getStETHByWstETH(amountIn);
+            }
+        }
+    }
+
+    function dexAmountOutFormatted(bytes memory payload, uint256 amountIn)
+        public
+        returns (uint256 amountOut, address midToken)
+    {
+        return dexAmountOut(payload, amountIn);
+    }
+
+    function dexAmountIn(bytes memory route, uint256 amountOut)
+        public
+        returns (uint256 amountIn, address midToken)
+    {
+        revert("unsupported");
+    }
+
+    function dexAmountInFormatted(bytes memory payload, uint256 amountOut)
+        public
+        returns (uint256 amountIn, address midToken)
+    {
+        revert("unsupported");
+    }
+
+    function setSwapApprovals(
+        address[] memory /*tokens*/
+    ) public {
+        IERC20(STETH).safeApprove(WSTETH, 0);
```

Comment:
safeApprove handling that

Reply:
safe approve throws an error when it is non-zero, it does not 0 it out for you